File: db/schema/spmb.ts
import { sqliteTable, text, integer, real, index } from "drizzle-orm/sqlite-core";
import { relations } from "drizzle-orm";
import { createId } from "@paralleldrive/cuid2";
import { users } from "./users";

// ==========================================
// SPMB Periods Table
// ==========================================

export const spmbPeriods = sqliteTable("spmb_periods", {
  id: text("id")
    .primaryKey()
    .$defaultFn(() => createId()),
  academicYear: text("academic_year").notNull(),
  name: text("name").notNull(),
  startDate: integer("start_date", { mode: "timestamp" }).notNull(),
  endDate: integer("end_date", { mode: "timestamp" }).notNull(),
  quota: integer("quota").default(0),
  isActive: integer("is_active", { mode: "boolean" }).default(false),
  createdAt: integer("created_at", { mode: "timestamp" }).$defaultFn(
    () => new Date()
  ),
  updatedAt: integer("updated_at", { mode: "timestamp" })
    .$defaultFn(() => new Date())
    .$onUpdate(() => new Date()),
});

// ==========================================
// SPMB Registrants Table
// ==========================================

export const registrantStatusEnum = [
  "draft",
  "pending",
  "verified",
  "accepted",
  "rejected",
] as const;

export const genderEnum = ["L", "P"] as const;

export const spmbRegistrants = sqliteTable(
  "spmb_registrants",
  {
    id: text("id")
      .primaryKey()
      .$defaultFn(() => createId()),
    registrationNumber: text("registration_number").unique().notNull(),
    // Student Details
    fullName: text("full_name").notNull(),
    studentName: text("student_name"), // Legacy/Duplicate
    nisn: text("nisn"), // Optional
    studentNik: text("student_nik").unique().notNull(),
    kkNumber: text("kk_number").notNull(),
    birthCertificateNo: text("birth_certificate_no"),
    birthDate: integer("birth_date", { mode: "timestamp" }).notNull(),
    birthPlace: text("birth_place").notNull(),
    gender: text("gender", { enum: genderEnum }).notNull(),
    religion: text("religion").notNull(),
    specialNeeds: text("special_needs"), // "Tidak" or specific need
    livingArrangement: text("living_arrangement"), // "Bersama Orang Tua", etc.
    transportMode: text("transport_mode"),
    childOrder: integer("child_order"),
    hasKpsPkh: integer("has_kps_pkh", { mode: "boolean" }).default(false),
    hasKip: integer("has_kip", { mode: "boolean" }).default(false),
    previousSchool: text("previous_school"),
    
    // Detailed Address
    addressStreet: text("address_street").notNull(),
    addressRt: text("address_rt").notNull(),
    addressRw: text("address_rw").notNull(),
    addressVillage: text("address_village").notNull(),
    postalCode: text("postal_code"),
    // Computed/Legacy address field for backward compat or summary
    address: text("address").notNull(),

    // Contact (Parent/Guardian Representative)
    parentName: text("parent_name").notNull().default("-"), // Legacy
    parentPhone: text("parent_phone").notNull(),
    parentEmail: text("parent_email"), // Optional

    // Father Details
    fatherName: text("father_name"),
    fatherNik: text("father_nik"),
    fatherBirthYear: text("father_birth_year"),
    fatherEducation: text("father_education"),
    fatherJob: text("father_job"),
    fatherIncome: text("father_income"),

    // Mother Details
    motherName: text("mother_name"),
    motherNik: text("mother_nik"),
    motherBirthYear: text("mother_birth_year"),
    motherEducation: text("mother_education"),
    motherJob: text("mother_job"),
    motherIncome: text("mother_income"),

    // Guardian Details
    guardianName: text("guardian_name"),
    guardianNik: text("guardian_nik"),
    guardianBirthYear: text("guardian_birth_year"),
    guardianEducation: text("guardian_education"),
    guardianJob: text("guardian_job"),
    guardianIncome: text("guardian_income"),

    homeLat: real("home_lat"),
    homeLng: real("home_lng"),
    distanceToSchool: real("distance_to_school"),
    isInZone: integer("is_in_zone", { mode: "boolean" }).default(false),
    status: text("status", { enum: registrantStatusEnum }).default("draft"),
    notes: text("notes"),
    priorityRank: integer("priority_rank"),
    priorityGroup: integer("priority_group"),
    periodId: text("period_id").references(() => spmbPeriods.id),
    documents: text("documents"), // JSON array of file paths
    verifiedBy: text("verified_by").references(() => users.id),
    verifiedAt: integer("verified_at", { mode: "timestamp" }),
    createdAt: integer("created_at", { mode: "timestamp" }).$defaultFn(
      () => new Date()
    ),
    updatedAt: integer("updated_at", { mode: "timestamp" })
      .$defaultFn(() => new Date())
      .$onUpdate(() => new Date()),
  },
  (table) => ({
    statusIdx: index("idx_registrant_status").on(table.status),
    createdIdx: index("idx_registrant_created").on(table.createdAt),
    periodIdx: index("idx_registrant_period").on(table.periodId),
  })
);

// ==========================================
// Relations
// ==========================================

export const spmbPeriodsRelations = relations(spmbPeriods, ({ many }) => ({
  registrants: many(spmbRegistrants),
}));

export const spmbRegistrantsRelations = relations(
  spmbRegistrants,
  ({ one }) => ({
    period: one(spmbPeriods, {
      fields: [spmbRegistrants.periodId],
      references: [spmbPeriods.id],
    }),
    verifier: one(users, {
      fields: [spmbRegistrants.verifiedBy],
      references: [users.id],
    }),
  })
);

// ==========================================
// Types
// ==========================================

export type SPMBPeriod = typeof spmbPeriods.$inferSelect;
export type NewSPMBPeriod = typeof spmbPeriods.$inferInsert;
export type SPMBRegistrant = typeof spmbRegistrants.$inferSelect;
export type NewSPMBRegistrant = typeof spmbRegistrants.$inferInsert;
export type RegistrantStatus = (typeof registrantStatusEnum)[number];
export type Gender = (typeof genderEnum)[number];

File: go-backend/internal/models/spmb.go
package models

import (
	"time"
)

type SPMBPeriod struct {
	ID            string     `json:"id"`
	Year          string     `json:"year"` // Formally school_year or year
	AcademicYear  string     `json:"academic_year"`
	Name          string     `json:"name"`
	CommitteeName string     `json:"committeeName"`
	StartDate     *time.Time `json:"startDate"`
	EndDate       *time.Time `json:"endDate"`
	Status        string     `json:"status"` // active, archived, draft
	IsActive      bool       `json:"isActive"`
	Quota         int        `json:"quota"`
	Registered    int        `json:"registered"`
	CreatedAt     *time.Time `json:"createdAt"`
}

type SPMBRegistrant struct {
	ID                 string `json:"id"`
	RegistrationNumber string `json:"registrationNumber"`
	FullName           string `json:"fullName"`
	NISN               string `json:"nisn"`
	StudentNIK         string `json:"studentNik"`
	KKNumber           string `json:"kkNumber"`
	BirthCertificateNo string `json:"birthCertificateNo"`
	BirthPlace         string `json:"birthPlace"`
	BirthDate          string `json:"birthDate"`
	Gender             string `json:"gender"`
	Religion           string `json:"religion"`
	SpecialNeeds       string `json:"specialNeeds"`
	LivingArrangement  string `json:"livingArrangement"`
	TransportMode      string `json:"transportMode"`
	ChildOrder         int    `json:"childOrder"`
	HasKPS             bool   `json:"hasKpsPkh"`
	HasKIP             bool   `json:"hasKip"`
	PreviousSchool     string `json:"previousSchool"`

	Hobby             string `json:"hobby"`
	Ambition          string `json:"ambition"`
	Height            int    `json:"height"`
	Weight            int    `json:"weight"`
	HeadCircumference int    `json:"headCircumference"`
	SiblingCount      int    `json:"siblingCount"`
	TravelTime        string `json:"travelTime"`

	AddressStreet  string  `json:"addressStreet"`
	AddressRT      string  `json:"addressRt"`
	AddressRW      string  `json:"addressRw"`
	AddressVillage string  `json:"addressVillage"`
	PostalCode     string  `json:"postalCode"`
	HomeAddress    string  `json:"address"`
	HomeLat        float64 `json:"homeLat"`
	HomeLng        float64 `json:"homeLng"`
	DistanceKM     float64 `json:"distanceToSchool"`
	IsInZone       bool    `json:"isWithinZone"`

	ParentPhone    string `json:"parentPhone"`
	ParentEmail    string `json:"parentEmail"`
	FatherName     string `json:"fatherName"`
	FatherNIK      string `json:"fatherNik"`
	FatherBirth    string `json:"fatherBirthYear"`
	FatherEdu      string `json:"fatherEducation"`
	FatherJob      string `json:"fatherJob"`
	FatherIncome   string `json:"fatherIncome"`
	MotherName     string `json:"motherName"`
	MotherNIK      string `json:"motherNik"`
	MotherBirth    string `json:"motherBirthYear"`
	MotherEdu      string `json:"motherEducation"`
	MotherJob      string `json:"motherJob"`
	MotherIncome   string `json:"motherIncome"`
	GuardianName   string `json:"guardianName"`
	GuardianNIK    string `json:"guardianNik"`
	GuardianBirth  string `json:"guardianBirthYear"`
	GuardianEdu    string `json:"guardianEducation"`
	GuardianJob    string `json:"guardianJob"`
	GuardianIncome string `json:"guardianIncome"`

	Status     string `json:"status"`
	Notes      string `json:"notes"`
	PeriodID   string `json:"periodId"`
	Documents  string `json:"documents"`
	VerifiedAt int64  `json:"verifiedAt"`
	VerifiedBy string `json:"verifiedBy"`
	CreatedAt  int64  `json:"createdAt"`
	UpdatedAt  int64  `json:"updatedAt"`
}

type SPMBStats struct {
	TotalRegistrants    int `json:"total"`
	VerifiedRegistrants int `json:"verified"`
	PendingRegistrants  int `json:"pending"`
	AcceptedRegistrants int `json:"accepted"`
	RejectedRegistrants int `json:"rejected"`
}

type PublicLandingData struct {
	SchoolName          string  `json:"school_name"`
	SchoolNPSN          string  `json:"school_npsn"`
	SchoolAddress       string  `json:"school_address"`
	SchoolPhone         string  `json:"school_phone"`
	SchoolEmail         string  `json:"school_email"`
	SchoolLat           float64 `json:"school_lat"`
	SchoolLng           float64 `json:"school_lng"`
	SPMBIsOpen          bool    `json:"spmb_is_open"`
	CurrentAcademicYear string  `json:"current_academic_year"`
	MaxDistanceKM       float64 `json:"max_distance_km"`
	PrincipalName       string  `json:"principal_name"`
	PrincipalNIP        string  `json:"principal_nip"`
}

type CreateSPMBPeriodRequest struct {
	Year               string `json:"year"`
	AcademicYear       string `json:"academic_year"`
	AcademicYearCamel  string `json:"academicYear"`
	Name               string `json:"name"`
	CommitteeName      string `json:"committeeName"`
	CommitteeNameSnake string `json:"committee_name"`
	StartDate          string `json:"startDate"`
	EndDate            string `json:"endDate"`
	Status             string `json:"status"`
	IsActive           bool   `json:"isActive"`
	Quota              int    `json:"quota"`
}

type UpdateSPMBPeriodRequest struct {
	Year               *string `json:"year"`
	AcademicYear       *string `json:"academic_year"`
	AcademicYearCamel  *string `json:"academicYear"`
	Name               *string `json:"name"`
	CommitteeName      *string `json:"committeeName"`
	CommitteeNameSnake *string `json:"committee_name"`
	StartDate          *string `json:"startDate"`
	EndDate            *string `json:"endDate"`
	Status             *string `json:"status"`
	IsActive           *bool   `json:"isActive"`
	Quota              *int    `json:"quota"`
}

type SPMBPromoteRequest struct {
	RegistrantID string `json:"registrant_id"`
	ClassID      string `json:"class_id"`
	StudentID    string `json:"student_id"` // Optional, if already exists
}
